### PR TITLE
ARROW-13736: [C++] Reconcile PrettyPrint and StringFormatter

### PR DESCRIPTION
**Pull request by pitrou**
Also fix propagation of printing options in nested arrays such as lists.

### CODE REVIEW

File: python/pyarrow/tests/test_table.py [hunk 1]
```diff
@@ -1625,8 +1625,8 @@ def test_table_repr_to_string_ellipsis():
 c0: int16
 c1: int32
 ----
-c0: [[1,2,3,4,1,2,3,4,1,2,...3,4,1,2,3,4,1,2,3,4]]
-c1: [[10,20,30,40,10,20,30,40,10,20,...30,40,10,20,30,40,10,20,30,40]]"""
+c0: [[1,2,3,4,1,2,3,4,1,2,...,3,4,1,2,3,4,1,2,3,4]]
+c1: [[10,20,30,40,10,20,30,40,10,20,...,30,40,10,20,30,40,10,20,30,40]]"""
```
Comment (pitrou) [inline python/pyarrow/tests/test_table.py:1629]: @amol- Just for the record.

File: cpp/src/arrow/pretty_print.cc [hunk 2]
```diff
@@ -54,17 +57,21 @@ class PrettyPrinter {
...
   void OpenArray(const Array& array);
   void CloseArray(const Array& array);
-
   void Flush() { (*sink_) << std::flush; }
 
+  PrettyPrintOptions ChildOptions() const {
+    PrettyPrintOptions child_options = options_;
+    child_options.indent = indent_;
+    return child_options;
+  }
```
Comment (lidavidm) [inline cpp/src/arrow/pretty_print.cc:73]: This fixes the case of a non-default indent_size not being propagated, right? Should we add a test for that?

**Review by lidavidm — APPROVED** [review]
LGTM.

Not really relevant here, but it seems Python binds indent and not indent_size, which may be confusing to people expecting something like the json module's indent.

**Comment by lidavidm** [comment]
> I think `indent_size` is just like the Python `json` module's indent. The names and descriptions are just a bit confusing.

Right, what I meant is that Python binds `indent`, hence it's a fixed indent for all lines, not an indent-per-item like people might expect. https://github.com/apache/arrow/blob/3fb1e7dd98f591c073052f6d8794185584702ca3/python/pyarrow/array.pxi#L999-L1032

**Comment by pitrou** [comment]
I see. Perhaps open a JIRA for that?

**Pull request closed without merging**